synapse/__init__.py

```diff
@@ -64,7 +64,9 @@
 try:
     from canonicaljson import set_json_library
 
-    set_json_library(json)
+    # type-ignore: I think there has been a regression in mypy 1.0.0 in how it checks
+    # modules against Protocols.
```

Comment:
python/mypy#13513 supported this, maybe it is buggy?

Reply:
I suspect it works 90% of the time, but fails in more complex/nested cases.

I've encountered pain here before when I was trying to define the `DBAPI2` Protocol:

https://github.com/matrix-org/synapse/blob/1fe202a1a3343fad77da270ffe0923a46f1944dd/synapse/storage/types.py#L118-L121

See also the commit message for the error that mypy spewed out: https://github.com/matrix-org/synapse/pull/15052/commits/408935c7f5c7b0393566ce2bc51e3ad98eb3a080

If I can find the time I'll try to reduce to a minimal test case and report a mypy bug.

Reply:
e9329dc7a1 and 029cfaf046 mean we don't have to suppress this error any more.